### ISSUE TITLE
Bilinear interpolation and Confidence thresholding

### ISSUE THREAD

**Issue by danielm322** [SamWilso/SAFE_Official#3]
Hello Dear Samuel, I al Daniel Montoya, working also on Object level OoD detection at CEA (France)

We have been quite interested in your paper SAFE. I have a few questions:

1. In your paper, section 3.2, subsection "Object-specific SAFE extraction"  you state that: "These object-specific feature maps O_{l,d} are then reduced to a vector representation p{l,d} for concatenation via a bilinear interpolation operation along the spatial axis".  However, when looking at your code, in the tracker.py script is where we suppose you are actually extracting the features, in the lines 45-52: 
```	def roi_features(self, rois, input_h):
		det_feats = []
		for feat in self.features:
			_, _, h, _ = feat.size()
			scale = h/input_h
			feat = roi_align(feat, rois, (1, 1), scale).mean(dim=(2, 3))
			det_feats.append(feat)
		return torch.cat(det_feats, dim=1)
```
what we can see is that the mean is extracted from each hooked roi, instead of a bilinear interpolation. Are we maybe not looking at the correct part of the code?

2. In section 4.1 you state that "Importantly, under the benchmark setting by [9], the AUROC and FPR95 metrics are computed after low confidence objects are suppressed according to a confidence threshold determined by [17]. Our comparisons in
Table 1 implement this suppression for fair comparisons (...)"  Where in the code is this thresholding operation implemented? Also since the description is not very detailed, how do you obtain this threshold? Is it the same threshold for InD samples and OoD samples?

Thank you for your answers!
Kind regards, 
Daniel

**Comment by SamWilso** [comment]
Hello Daniel,

Thank you for you interest in our work! I am happy to answer your questions:

1. The bilinear interpolation of the features is done as part of the `roi_align(...)` operation as defined by the MaskRCNN paper. The third argument passed in `roi_align(feat, rois, (1, 1), scale)` (i.e. the (1, 1) tuple) defines the output size of the feature with the `.mean(dim=(2,3))` serving to simply remove the remaining 1x1 (HxW) dimensions for easier processing.

2. This threshold is determined as part of the VOS benchmark and is computed automatically in `offline_evaluation/compute_average_precision.py`. Since it a part of the benchmark itself and not our work, we did not define it in the conference submission.
It requires a bit of digging through the code of VOS to figure out, but in short: The confidence threshold is determined by finding the value which maximises the mAP on the ID data. After this, all detections (ID or OOD) below this confidence threshold will be suppressed and excluded from the OOD detection computation. In the original VOS submission, this thresholding operation is applied to *all* OOD detectors evaluated. 
As this threshold is solely dependant on the ID-task performance of the object detector, our post-hoc SAFE detector does not affect this computation.

Please let me know if I can help any further.
